hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeysRenameRequest.java

```diff
@@ -0,0 +1,298 @@
+            //                   the request would be deemed a replay. But
+            //                   Key1 is still in the DB and needs to be
+            //                   deleted.
+            fromKeyValue = omMetadataManager.getKeyTable().get(fromKey);
+            if (fromKeyValue != null) {
+              // Check if this replay transaction was after the fromKey was
+              // created. If so, we have to delete the fromKey.
+              if (ozoneManager.isRatisEnabled() &&
+                  trxnLogIndex > fromKeyValue.getUpdateID()) {
+                // Add to cache. Only fromKey should be deleted. ToKey already
+                // exists in DB as this transaction is a replay.
+                result = Result.DELETE_FROM_KEY_ONLY;
+                renameKeyInfoList.add(new OmRenameKeyInfo(
+                    null, fromKeyValue));
+              }
+            }
+
+            if (result == null) {
+              result = Result.REPLAY;
+              // If toKey exists and fromKey does not, then no further action is
+              // required. Return a dummy OMClientResponse.
+              omClientResponse =
+                  new OMKeysRenameResponse(createReplayOMResponse(
+                      omResponse));
+            }
+          } else {
+            // This transaction is not a replay. toKeyName should not exist
+            throw new OMException("Key already exists " + toKeyName,
+                OMException.ResultCodes.KEY_ALREADY_EXISTS);
+          }
+        } else {
+          // fromKeyName should exist
+          fromKeyValue = omMetadataManager.getKeyTable().get(fromKey);
+          if (fromKeyValue == null) {
+            // TODO: Add support for renaming open key
+            throw new OMException("Key not found " + fromKey, KEY_NOT_FOUND);
+          }
+
+          fromKeyValue.setUpdateID(trxnLogIndex, ozoneManager.isRatisEnabled());
+
+          fromKeyValue.setKeyName(toKeyName);
+          //Set modification time
+          fromKeyValue.setModificationTime(renameKeyArgs.getModificationTime());
+
+          renameKeyInfoList
+              .add(new OmRenameKeyInfo(fromKeyName, fromKeyValue));
+        }
+      }
+      omClientResponse = new OMKeysRenameResponse(omResponse
+          .setRenameKeysResponse(RenameKeysResponse.newBuilder()).build(),
+          renameKeyInfoList, trxnLogIndex);
+      result = Result.SUCCESS;
+    } catch (IOException ex) {
+      result = Result.FAILURE;
+      exception = ex;
+      omClientResponse = new OMKeyDeleteResponse(
+          createRenameKeysErrorOMResponse(omResponse, exception,
+              unRenamedKeys));
+    } finally {
+      addResponseToDoubleBuffer(trxnLogIndex, omClientResponse,
+          omDoubleBufferHelper);
+    }
+
+    if (result == Result.SUCCESS || result == Result.FAILURE) {
+      auditLog(auditLogger, buildAuditMessage(OMAction.RENAME_KEY, auditMap,
+          exception, getOmRequest().getUserInfo()));
```

Comment:
Shouldn't all renamed keys be logged to audit?

Reply:
Thanks @adoroszlai  for the review. Fixed this issue.

Reply:
Thanks, but the new code only logs key names, other info eg. volume/bucket is missing.  Also, if result is failure, then all keys are logged as "failure", although only the last key failed (and we don't even know which one is last, since `auditMap` is a hashmap).

Reply:
Thanks adoroszlai  for the tip，I'll add a volume/bucket to the key. 
Also, on how to locate the failed key. We can see this in the Exception in auditLog. Typically, the failed keys are displayed in the log with an Exception, such as:
`17:45:30.550 [OM StateMachine ApplyTransaction Thread - 0] ERROR OMAudit - user=micahzhao | ip=127.0.0.1 | op=RENAME_KEY {dir/key2=dir/file2, dir/file1=dir/file2} | ret=FAILURE`
`org.apache.hadoop.ozone.om.exceptions.OMException: Key not found /a88fb570-5b5d-43db-81e0-d6597f4ea81f/4ad1e6c3-41c3-439d-8082-ae24a601ba38/dir/file1`
`at org.apache.hadoop.ozone.om.request.key.OMKeysRenameRequest.validateAndUpdateCache(OMKeysRenameRequest.java:228)`